-*- mode: org -*-
#+STARTUP: overview
#+STARTUP: hideblocks
#+STARTUP: hidestars

* Timeline
** TODO dmr: 
  (1) one test vs. one control: current implementation seems not quite
   easy to use; lack validation, evaluation and comparison with other
   methods; 1 week (2) lack support for experiment design with
   replicates: time undefined

** DONE hmr_plant: Viterbi decoding:
   CLOSED: [2012-11-26 Mon 16:22]
    to avoid inconsistent sequences for Posterior decoding: 3 days  

** TODO hmr

   with small test dataset, hmr don't output any hmr after filtering
   by false discovered rate
   
** DONE pair-end support for rmapbs
   CLOSED: [2012-12-04 Tue 20:21] SCHEDULED: <2012-11-06 Tue>

   need to address memory issue and speed issue

   rmapbs-pe: input: T-rich and A-rich reads input; output: clipped
   DNA fragments as the output from clipmates

** DONE test dataset 
   CLOSED: [2012-11-26 Mon 16:26]
   (1) need more coverage: current test dataset does not have enough
   coverage; (2) need two samples to test DMR program; (3) make sure
   the selected regions also contain a few AMRs 

   /home/rcf-47/qiangson/panfs/methpipe/data   

** DONE adapt methcounts and other programs to use new format
   CLOSED: [2012-11-16 Fri 20:58] DEADLINE: <2012-11-13 Tue> SCHEDULED: <2012-11-11 Sun>

methcounts output:
chrom location strand [seq] meth coverage 
chr1  108      +      CpG   0.2          5

file format for multiple samples: in two files: one contains
methylation level and the other contains coverage

*methylation* 
chrom location strand seq  sample1 sample2 sample3
chr1  108   + CpG 0.2 0.5 0.6


*coverage* 
chrom location strand seq  sample1 sample2 sample3
chr1  108   + CpG 8 9 10

** DONE test dataset  
   CLOSED: [2012-11-07 Wed 11:22] DEADLINE: <2012-11-06 Tue> SCHEDULED: <2012-11-06 Tue>
   - need more coverage: current test dataset does not have enough
     coverage;
   - need two samples to test DMR program;
   - make sure the selected regions also contain a few AMRs
   - RRBS dataset

* TODOs 

- development: 

  1.  dmr: one test vs. one control: current implementation seems not
      quite easy to use; lack validation, evaluation and comparison
      with other methods; 1 week (2) lack support for experiment
      design with replicates: time undefined
  
  2. hmr_plant: Viterbi decoding:

  3. hmr: with small test dataset, hmr don't output any hmr after
     filtering by false discovered rate
   
  4. pair-end support for rmapbs

  5. adapt methcounts and other programs to use new format

  3. wrapper for other mappers in case users want other faster tools:
     BSMAP (implemented and used in the Arabidopsis project); Bismark;   

  5. dmr: (1) one test vs. one control: current implementation seems
     not quite easy to use; lack validation, evaluation and comparison
     with other methods; 1 week (2) lack support for experiment design
     with replicates: time undefined

  6. amr: merge from separate program to methpipe package: 1day
	 
  7. adapt the pipeline to galaxy platform (long term goal)
	 
- testing, documentation and evaluation 
  
  1. test compiling on Linux and Mac OS: 1 day 
  2. proof-read the manual, update and refine 
  4. use makefile to automate the above test results

- RRBS related
  1. sanity check: run methpipe and those tools recommended from the
     original papers on some RRBS dataset
  2. mapper: have not looked carefully, but the RRBS proctocol paper
     stress the specialties in dealing with RRBS reads, including (1)
     adapter trimming; and (2) any RRBS reads should have one end
     located at CCGG, which may help reduce incorrectly or ambiguous
     mapped reads  
  3. collect several RRBS dataset and check how consistent the
     enriched regions are between experiments;
  4. compare how well the methylation estimates from RRBS and WGBS
     are correlated (require some new experiment from RRBS and
     WGBS. push someone run both RRBS and WGBS over two different
     samples to evaluate how RRBS can recover DMRs identified from
     WGBS. RRBS represents very poorly of CGI shores and enhancers)
  5. run HMR program on selected RRBS dataset. Since RRBS profiles
     those CpG sites close to CCGG motif, it is possible that this is
     no continuous regions at all.
